File: AjaxVectorObjects/ClientScripts/RedoUndo/HistoryData.js
// Copyright (c) 2018 Aurigma Inc. All rights reserved.
// Licensed under the MIT License. See LICENSE file in the project root for full license information.
//
Type.registerNamespace("Aurigma.GraphicsMill.AjaxControls.VectorObjects.RedoUndo");

Aurigma.GraphicsMill.AjaxControls.VectorObjects.RedoUndo.HistoryData = function (history) {
    ///	<summary>This class represents a state of a history and allows applying this state to the <see cref="T:J:Aurigma.GraphicsMill.AjaxControls.VectorObjects.RedoUndo.History" /> class instance.</summary>
    /// <constructor>
    /// 	<exclude />
    /// </constructor>
    this.Cur = -1;
    this.E = false;
    this.L = false;
    this.US = 10;
    this.T = true;
    this.C = [];
    if (history) {
        this.Cur = history.get_current();
        this.E = history.get_enable();
        this.L = history.get_locked();
        this.US = history.get_maxUndoStepCount();
        this.T = history.get_trackingEnabled();
        this.C = [];
        var commands = history.get_commands();
        for (var i = 0; i < commands.length; i++) {
            var command = commands[i];
            var type = Object.getType(command).getName();
            var data = "";
            if (type == "Aurigma.GraphicsMill.AjaxControls.VectorObjects.RedoUndo.VObjectGroupCommand" && !command.isEmpty()) {
                var groupCommand = {};
                groupCommand._commands = [];
                for (var j = 0; j < command.get_commands().length; j++) {
                    var c = command.get_commands()[j];
                    var t = Object.getType(c).getName();
                    var d = JSON.stringify(c);
                    groupCommand._commands.push({ D: d, T: t });
                }
                data = JSON.stringify(groupCommand);
            }
            else {
                data = JSON.stringify(commands[i]);
            }

            this.C.push({ D: data, T: type });
        }
    }
}

Aurigma.GraphicsMill.AjaxControls.VectorObjects.RedoUndo.HistoryData.applyState = function (historyData, history) {
    /// <summary>Applies the <paramref name="historyData" /> state to the specified <see cref="T:J:Aurigma.GraphicsMill.AjaxControls.VectorObjects.RedoUndo.History" />.</summary>
    /// <param name="historyData" type="Aurigma.GraphicsMill.AjaxControls.VectorObjects.RedoUndo.HistoryData">The state to apply.</param>
    /// <param name="history" type="Aurigma.GraphicsMill.AjaxControls.VectorObjects.RedoUndo.HistoryData">History to apply the state to.</param>
    if (!historyData) {
        return;
    }
    history.set_current(historyData.Cur);
    history.set_enable(historyData.E);
    history.set_locked(historyData.L);
    history.set_maxUndoStepCount(historyData.US);
    history.set_trackingEnabled(historyData.T);
    history.set_commands([]);
    for (var i = 0; i < historyData.C.length; i++) {
        if (historyData.C[i].D) {
            var type = historyData.C[i].T;
            var data = JSON.parse(historyData.C[i].D);
            var command = Aurigma.GraphicsMill.AjaxControls.VectorObjects.ObjectFactory.createObjectByType(type);

            if (type == "Aurigma.GraphicsMill.AjaxControls.VectorObjects.RedoUndo.VObjectGroupCommand" && data._commands.length > 0) {
                for (var j = 0; j < data._commands.length; j++) {
                    var t = data._commands[j].T;
                    var d = JSON.parse(data._commands[j].D);
                    var c = Aurigma.GraphicsMill.AjaxControls.VectorObjects.ObjectFactory.createObjectByType(t);
                    for (memberName in d) {
                        c[memberName] = d[memberName];
                    }
                    command.addCommand(c);
                }
            }
            else {
                for (memberName in data) {
                    command[memberName] = data[memberName];
                }
            }

            history.get_commands().push(command);
        }
    }
}

Aurigma.GraphicsMill.AjaxControls.VectorObjects.RedoUndo.HistoryData.registerClass("Aurigma.GraphicsMill.AjaxControls.VectorObjects.RedoUndo.HistoryData", null);

File: AjaxVectorObjects/Common/RedoUndo/HistoryData.cs
// Copyright (c) 2018 Aurigma Inc. All rights reserved.
// Licensed under the MIT License. See LICENSE file in the project root for full license information.
//
using System;
using System.Collections.Generic;

namespace Aurigma.GraphicsMill.AjaxControls.VectorObjects.RedoUndo
{
    [Serializable]
    public class HistoryTypeData
    {
        public HistoryTypeData()
            : this(null, null)
        { }

        public HistoryTypeData(string type, string data)
        {
            this.T = type;
            this.D = data;
        }

        public string T;
        public string D;
    }

    [Serializable]
    public class HistoryData
    {
        public HistoryData()
        {
            Cur = -1;
            E = false;
            L = false;
            T = true;
            US = 10;
            C = new List<HistoryTypeData>();
        }

        public HistoryData(History history)
        {
            Cur = history.Current;
            E = history.Enable;
            L = history.Locked;
            T = history.TrackingEnabled;
            US = history.MaxUndoStepCount;
            if (history.Commands != null)
            {
                C = new List<HistoryTypeData>();

                var serializer = new JsonVOSerializer();
                foreach (var command in history.Commands)
                {
                    HistoryTypeData data = null;
                    var unknownCommand = command as UnknownCommand;
                    if (unknownCommand != null)
                    {
                        data = new HistoryTypeData(unknownCommand.Type, unknownCommand.Data);
                    }
                    else if (command != null)
                    {
                        data = new HistoryTypeData(command.GetType().FullName,
                            serializer.Serialize(command));
                    }

                    if (data != null)
                    {
                        C.Add(data);
                    }
                }
            }
            else
            {
                C = new List<HistoryTypeData>();
            }
        }

        public void ApplyState(History history)
        {
            history.Current = Cur;
            history.Enable = E;
            history.Locked = L;
            history.MaxUndoStepCount = US;
            history.Commands.Clear();
            history.TrackingEnabled = T;
            for (int i = 0; i < C.Count; i++)
            {
                if (!String.IsNullOrEmpty(C[i].D))
                {
                    var serializer = new JsonVOSerializer();

                    Command command = null;
                    switch (C[i].T)
                    {
                        case "Aurigma.GraphicsMill.AjaxControls.VectorObjects.RedoUndo.LayerAddedCommand":
                            command = serializer.Deserialize<RedoUndo.LayerAddedCommand>(C[i].D);
                            break;

                        case "Aurigma.GraphicsMill.AjaxControls.VectorObjects.RedoUndo.LayerRemovedCommand":
                            command = serializer.Deserialize<RedoUndo.LayerRemovedCommand>(C[i].D);
                            break;

                        case "Aurigma.GraphicsMill.AjaxControls.VectorObjects.RedoUndo.LayerMovedCommand":
                            command = serializer.Deserialize<RedoUndo.LayerMovedCommand>(C[i].D);
                            break;

                        case "Aurigma.GraphicsMill.AjaxControls.VectorObjects.RedoUndo.VObjectAddedCommand":
                            command = serializer.Deserialize<RedoUndo.VObjectAddedCommand>(C[i].D);
                            break;

                        case "Aurigma.GraphicsMill.AjaxControls.VectorObjects.RedoUndo.VObjectChangedCommand":
                            command = serializer.Deserialize<RedoUndo.VObjectChangedCommand>(C[i].D);
                            break;

                        case "Aurigma.GraphicsMill.AjaxControls.VectorObjects.RedoUndo.VObjectRemovedCommand":
                            command = serializer.Deserialize<RedoUndo.VObjectRemovedCommand>(C[i].D);
                            break;

                        case "Aurigma.GraphicsMill.AjaxControls.VectorObjects.RedoUndo.VObjectMovedCommand":
                            command = serializer.Deserialize<RedoUndo.VObjectMovedCommand>(C[i].D);
                            break;

                        default:
                            command = new UnknownCommand() { Type = C[i].T, Data = C[i].D };
                            break;
                    }

                    if (command != null)
                    {
                        history.Commands.Add(command);
                    }
                }
            }
        }

        /// <summary>
        /// TrackingEnabled
        /// </summary>
        public bool T { get; set; }

        /// <summary>
        /// MaxUndoStepCount
        /// </summary>
        public int US { get; set; }

        /// <summary>
        /// Commands
        /// </summary>
        public List<HistoryTypeData> C { get; set; }

        /// <summary>
        /// Enable
        /// </summary>
        public bool E { get; set; }

        /// <summary>
        /// Locked
        /// </summary>
        public bool L { get; set; }

        /// <summary>
        /// Current
        /// </summary>
        public int Cur { get; set; }
    }
}